org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/JavaCodeMiningReconciler.java

```diff
@@ -47,15 +48,22 @@ public class JavaCodeMiningReconciler implements IJavaReconcilingListener {
 	public void reconciled(CompilationUnit ast, boolean forced, IProgressMonitor progressMonitor) {
 		final ISourceViewerExtension5 sourceViewer= fSourceViewer; // take a copy as this can be null-ed in the meantime
 		if (sourceViewer != null) {
-			reconciledViewers.add(sourceViewer);
+			CompletableFuture<CompilationUnit> future= toBeReconciledViewers.remove(sourceViewer);
+			if (future!=null) {
+				future.complete(ast);
+			}
 			sourceViewer.updateCodeMinings();
 		}
 	}
 
+	public static CompletableFuture<CompilationUnit> getFuture(ISourceViewerExtension5 viewer) {
+		CompletableFuture<CompilationUnit> future= toBeReconciledViewers.get(viewer);
+		return future;
+	}
+
 	@Override
 	public void aboutToBeReconciled() {
-		// interrupt code minings if modification occurs
-		reconciledViewers.remove(fSourceViewer);
+		toBeReconciledViewers.computeIfAbsent(fSourceViewer, unused->{return new CompletableFuture<CompilationUnit>();});
```

Comment:
I would more would use compute here and then check if there is already one (and cancel it) and creating always a fresh one.